server/src/main/java/org/elasticsearch/node/ShutdownPrepareService.java

```diff
@@ -0,0 +1,167 @@
+ */
+public class ShutdownPrepareService {
+
+    private final Logger logger = LogManager.getLogger(ShutdownPrepareService.class);
+    Settings settings;
+    HttpServerTransport httpServerTransport;
+    TerminationHandler terminationHandler;
+    private volatile boolean hasBeenShutdown = false;
+
+    public ShutdownPrepareService(Settings settings, HttpServerTransport httpServerTransport, TerminationHandler terminationHandler) {
+        this.settings = settings;
+        this.httpServerTransport = httpServerTransport;
+        this.terminationHandler = terminationHandler;
+    }
+
+    /**
+     * Invokes hooks to prepare this node to be closed. This should be called when Elasticsearch receives a request to shut down
+     * gracefully from the underlying operating system, before system resources are closed. This method will block
+     * until the node is ready to shut down.
+     * <p>
+     * Note that this class is part of infrastructure to react to signals from the operating system - most graceful shutdown
+     * logic should use Node Shutdown, see {@link org.elasticsearch.cluster.metadata.NodesShutdownMetadata}.
+     */
+    public void prepareForShutdown(TaskManager taskManager) {
+        assert hasBeenShutdown == false;
+        hasBeenShutdown = true;
+        final var maxTimeout = Node.MAXIMUM_SHUTDOWN_TIMEOUT_SETTING.get(settings);
+        final var reindexTimeout = Node.MAXIMUM_REINDEXING_TIMEOUT_SETTING.get(settings);
+
+        record Stopper(String name, SubscribableListener<Void> listener) {
+            boolean isIncomplete() {
+                return listener().isDone() == false;
+            }
+        }
+
+        final var stoppers = new ArrayList<Stopper>();
+        final var allStoppersFuture = new PlainActionFuture<Void>();
+        try (var listeners = new RefCountingListener(allStoppersFuture)) {
+            final BiConsumer<String, Runnable> stopperRunner = (name, action) -> {
+                final var stopper = new Stopper(name, new SubscribableListener<>());
+                stoppers.add(stopper);
+                stopper.listener().addListener(listeners.acquire());
+                new Thread(() -> {
+                    try {
+                        action.run();
+                    } catch (Exception ex) {
+                        logger.warn("unexpected exception in shutdown task [" + stopper.name() + "]", ex);
+                    } finally {
+                        stopper.listener().onResponse(null);
+                    }
+                }, stopper.name()).start();
+            };
+
+            stopperRunner.accept("http-server-transport-stop", httpServerTransport::close);
+            stopperRunner.accept("async-search-stop", () -> awaitSearchTasksComplete(maxTimeout, taskManager));
+            stopperRunner.accept("reindex-stop", () -> awaitReindexTasksComplete(reindexTimeout, taskManager));
```

Comment:
Is this the only change in this class? The refactor makes it a bit harder to reason about the change. We can keep it as is, but it might be beneficial to do the refactor separately for future PRs.

Reply:
The refactor was done for testing purpose as explained above.

Reply:
Yeah, I understand that. Still, is the line added (plus the method called) the only change here or has other changes sneaked in?

Reply:
Oh I see, I thought your question was about creating this new class. I only really added the awaitReindexTasksComplete() but in order to reuse the code being used by awaitSearchTasksComplete I created a new API waitForTasks() to replace [awaitSearchTasksComplete()](https://github.com/elastic/elasticsearch/blob/98d53352160d6d2c397ea74567ebc690b701973f/server/src/main/java/org/elasticsearch/node/Node.java#L664)